src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProjectFactory.cs

```diff
@@ -229,86 +249,101 @@ public void ApplyChangeToWorkspace(ProjectId projectId, Func<CodeAnalysis.Soluti
-                    return solutionChanges.Solution;
-                },
-                changeKind: (_, _) => (solutionChanges.WorkspaceChangeKind, solutionChanges.WorkspaceChangeProjectId, solutionChanges.WorkspaceChangeDocumentId),
-                onBeforeUpdate: (_, _) =>
-                {
-                    // Clear out mutable state not associated with the solution snapshot (for example, which documents are
-                    // currently open).
-                    foreach (var documentId in solutionChanges.DocumentIdsRemoved)
-                        Workspace.ClearDocumentData(documentId);
-                },
-                onAfterUpdate: null,
-                CancellationToken.None).ConfigureAwait(false);
+            await ApplyBatchChangeToWorkspaceMaybe_NoLockAsync(useAsync, mutation, onAfterUpdate).ConfigureAwait(false);
         }
     }
 
-    private void ApplyBatchChangeToWorkspace_NoLock(SolutionChangeAccumulator solutionChanges)
+    /// <summary>
+    /// Applies a change to the workspace that can do any number of project changes.
+    /// The mutation action must be safe to attempt multiple times, in case there are interceding solution changes.
+    /// If outside changes need to run under the global lock and run only once, they should use the <paramref name="onAfterUpdate"/> action.
+    /// </summary>
+    /// <remarks>This is needed to synchronize with <see cref="ApplyChangeToWorkspace(Action{Workspace})" /> to avoid any races. This
+    /// method could be moved down to the core Workspace layer and then could use the synchronization lock there.</remarks>
+    public async Task ApplyBatchChangeToWorkspaceMaybe_NoLockAsync(bool useAsync, Func<SolutionChangeAccumulator, ProjectUpdateState, ProjectUpdateState> mutation, Action<ProjectUpdateState>? onAfterUpdate)
     {
-        Contract.ThrowIfFalse(_gate.CurrentCount == 0);
+        // We need the data from the accumulator across the lambda callbacks to SetCurrentSolutionAsync, so declare
+        // it here. It will be assigned in `transformation:` below (which may happen multiple times if the
+        // transformation needs to rerun).  Once the transformation succeeds and is applied, the
+        // 'onBeforeUpdate/onAfterUpdate' callbacks will be called, and can use the last assigned value in
+        // `transformation`.
+        SolutionChangeAccumulator solutionChanges = null!;
+        ProjectUpdateState projectUpdateState = null!;
+
+        var (didUpdate, newSolution) = await Workspace.SetCurrentSolutionAsync(
+            useAsync,
+            transformation: oldSolution =>
+            {
+                solutionChanges = new SolutionChangeAccumulator(oldSolution);
 
-        if (!solutionChanges.HasChange)
-            return;
+                // Reset the project update state in case this is a retry (we don't want to save the results of the last attempt).
+                projectUpdateState = _projectUpdateState;
```

Comment:
this assignment seems unnecessary/undesirable.  Can just be merged into line below as: